<!DOCTYPE html>
<html lang="en" itemscope itemtype="http://schema.org/WebPage">
  <head>
    

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0">

  <title>Exact string matching algorithms: Boyer-Moore and KMP - Fan&#39;s Blog</title>
  <meta name="description" content="String matching algorithms are used in lots of scenarios such as searching words in a text file, or locating specific sequences in a genome. I&rsquo;ve heard of KMP algorithm long time ago, but don&rsquo;t have a chance to implement it by myself (just lazy). It seems that KMP is not widely used in genomics field, but instead another algorithm, Boyer-Moore, is taught more often in computational genomics classes. I recently find a great online tutorial on computational genomics by Ben Langmead @JHU (link).">
  <meta name="author" content="Fan"/><script type="application/ld+json">
{
    "@context": "http://schema.org",
    "@type": "WebSite",
    "name": "Fan\x27s Blog",
    
    "url": "https:\/\/dearxxj.github.io"
}
</script><script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Organization",
  "name": "",
  "url": "https:\/\/dearxxj.github.io"
  
  
  
  
}
</script>
<script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [{
        "@type": "ListItem",
        "position": 1,
        "item": {
          "@id": "https:\/\/dearxxj.github.io",
          "name": "home"
        }
    },{
        "@type": "ListItem",
        "position": 3,
        "item": {
          "@id": "https:\/\/dearxxj.github.io\/post\/4\/",
          "name": "Exact string matching algorithms boyer moore and k m p"
        }
    }]
}
</script><script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Article",
  "author": {
    "name" : "Fan"
  },
  "headline": "Exact string matching algorithms: Boyer-Moore and KMP",
  "description" : "String matching algorithms are used in lots of scenarios such as searching words in a text file, or locating specific sequences in a genome. I\x26rsquo;ve heard of KMP algorithm long time ago, but don\x26rsquo;t have a chance to implement it by myself (just lazy). It seems that KMP is not widely used in genomics field, but instead another algorithm, Boyer-Moore, is taught more often in computational genomics classes. I recently find a great online tutorial on computational genomics by Ben Langmead @JHU (link).",
  "inLanguage" : "en",
  "wordCount":  1310 ,
  "datePublished" : "2018-12-28T20:42:32",
  "dateModified" : "2018-12-28T20:42:32",
  "image" : "https:\/\/dearxxj.github.io\/img\/blog-log.png",
  "keywords" : [ "Algorithm, Bioinfo, Sequencing" ],
  "mainEntityOfPage" : "https:\/\/dearxxj.github.io\/post\/4\/",
  "publisher" : {
    "@type": "Organization",
    "name" : "https:\/\/dearxxj.github.io",
    "logo" : {
        "@type" : "ImageObject",
        "url" : "https:\/\/dearxxj.github.io\/img\/blog-log.png",
        "height" :  60 ,
        "width" :  60
    }
  }
}
</script>

<meta property="og:title" content="Exact string matching algorithms: Boyer-Moore and KMP" />
<meta property="og:description" content="String matching algorithms are used in lots of scenarios such as searching words in a text file, or locating specific sequences in a genome. I&rsquo;ve heard of KMP algorithm long time ago, but don&rsquo;t have a chance to implement it by myself (just lazy). It seems that KMP is not widely used in genomics field, but instead another algorithm, Boyer-Moore, is taught more often in computational genomics classes. I recently find a great online tutorial on computational genomics by Ben Langmead @JHU (link).">
<meta property="og:image" content="https://dearxxj.github.io/img/blog-log.png" />
<meta property="og:url" content="https://dearxxj.github.io/post/4/" />
<meta property="og:type" content="website" />
<meta property="og:site_name" content="Fan&#39;s Blog" />

  <meta name="twitter:title" content="Exact string matching algorithms: Boyer-Moore and KMP" />
  <meta name="twitter:description" content="String matching algorithms are used in lots of scenarios such as searching words in a text file, or locating specific sequences in a genome. I&rsquo;ve heard of KMP algorithm long time ago, but â€¦">
  <meta name="twitter:image" content="https://dearxxj.github.io/img/blog-log.png" />
  <meta name="twitter:card" content="summary" />
  <link href='https://dearxxj.github.io/img/favicon.ico' rel='icon' type='image/x-icon'/>
  <meta name="generator" content="Hugo 0.59.1" />
  <link rel="alternate" href="https://dearxxj.github.io/index.xml" type="application/rss+xml" title="Fan&#39;s Blog"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.css" integrity="sha384-9eLZqc9ds8eNjO3TmqPeYcDj8n+Qfa4nuSiGYa6DjLNcv9BtN69ZIulL9+8CqC9Y" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.5.0/css/all.css" integrity="sha384-B4dIYHKNBt8Bc12p+WXckhzcICo0wtJAoU8YZTY5qE0Id1GSseTk6S+L3BlXeVIU" crossorigin="anonymous">
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><link rel="stylesheet" href="https://dearxxj.github.io/css/main.css" /><link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lora:400,700,400italic,700italic" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,600italic,700italic,800italic,400,300,600,700,800" />
  <link rel="stylesheet" href="https://dearxxj.github.io/css/highlight.min.css" /><link rel="stylesheet" href="https://dearxxj.github.io/css/codeblock.css" /><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe.min.css" integrity="sha384-h/L2W9KefUClHWaty3SLE5F/qvc4djlyR4qY3NUV5HGQBBW7stbcfff1+I/vmsHh" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/default-skin/default-skin.min.css" integrity="sha384-iD0dNku6PYSIQLyfTOpB06F2KCZJAKLOThS5HRe8b3ibhdEQ6eKsFf/EeFxdOt5R" crossorigin="anonymous">


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-104430016-1', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>


  </head>
  <body>
    <nav class="navbar navbar-default navbar-fixed-top navbar-custom">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#main-navbar">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="https://dearxxj.github.io">Fan&#39;s Blog</a>
    </div>

    <div class="collapse navbar-collapse" id="main-navbar">
      <ul class="nav navbar-nav navbar-right">
        
          
            <li>
              <a title="About" href="/page/about/">About</a>
            </li>
          
        
          
            <li>
              <a title="Tags" href="/tags">Tags</a>
            </li>
          
        

        

        
      </ul>
    </div>

    
      <div class="avatar-container">
        <div class="avatar-img-border">
          <a title="Fan&#39;s Blog" href="https://dearxxj.github.io">
            <img class="avatar-img" src="https://dearxxj.github.io/img/blog-log.png" alt="Fan&#39;s Blog" />
          </a>
        </div>
      </div>
    

  </div>
</nav>




    


<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

<div class="pswp__bg"></div>

<div class="pswp__scroll-wrap">
    
    <div class="pswp__container">
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
    </div>
    
    <div class="pswp__ui pswp__ui--hidden">
    <div class="pswp__top-bar">
      
      <div class="pswp__counter"></div>
      <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>
      <button class="pswp__button pswp__button--share" title="Share"></button>
      <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>
      <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>
      
      
      <div class="pswp__preloader">
        <div class="pswp__preloader__icn">
          <div class="pswp__preloader__cut">
            <div class="pswp__preloader__donut"></div>
          </div>
        </div>
      </div>
    </div>
    <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
      <div class="pswp__share-tooltip"></div>
    </div>
    <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
    </button>
    <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
    </button>
    <div class="pswp__caption">
      <div class="pswp__caption__center"></div>
    </div>
    </div>
    </div>
</div>


  
  
  






  

  <header class="header-section ">
    
    <div class="intro-header no-img">
      <div class="container">
        <div class="row">
          <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
            <div class="post-heading">
              
                <h1>Exact string matching algorithms: Boyer-Moore and KMP</h1>
              
              
              
              
                <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on December 28, 2018
  
  
    &nbsp;|&nbsp;<i class="fas fa-clock"></i>&nbsp;7&nbsp;minutes
  
  
    &nbsp;|&nbsp;<i class="fas fa-book"></i>&nbsp;1310&nbsp;words
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;Fan
    
  
  
</span>


              
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>


    
<div class="container" role="main">
  <div class="row">
    <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
      <article role="main" class="blog-post">
        

<p>String matching algorithms are used in lots of scenarios such as searching words in a text file, or locating specific sequences in a genome. I&rsquo;ve heard of KMP algorithm long time ago, but don&rsquo;t have a chance to implement it by myself (<em>just lazy</em>). It seems that KMP is not widely used in genomics field, but instead another algorithm, Boyer-Moore, is taught more often in computational genomics classes. I recently find a great online tutorial on computational genomics by Ben Langmead @JHU (<a href="http://www.langmead-lab.org/teaching-materials/">link</a>). In the string exact matching section, Ben talks about the idea behind Boyer-Moore and kindly provides a python implementation which was based on the Z algorithm and <a href="https://www.cambridge.org/core/books/algorithms-on-strings-trees-and-sequences/F0B095049C7E6EF5356F0A26686C20D3">Dan Gusfield&rsquo;s book</a>. Like other teaching materials, Dan&rsquo;s book first looked at the naive approach for exact matching. But then this book went straightly into the Z algorithm, with which it nicely linked Boyer-Moore and KMP algorithms. For the first time, I know that both algorithms can be derived from the Z algorithm.</p>

<p>What interests me more is that this book mentions that <strong>KMP, while widely-known, is <em>much inferior</em> to the Boyer-Moore algorithm in practice</strong>. I can&rsquo;t help asking myself: Is that really the case? This question motivates me to run some tests on the two algorithms and write this article.</p>

<p>First, let&rsquo;s briefly go over the two string-matching algorithms. Unlike the naive algorithm which takes O(N*M) time, Boyer-Moore and KMP require only linear time to find all matches of a pattern in a string. What makes Boyer-Moore and KMP fast is the pre-processing of the pattern, which can be used to skip characters during the comparison. The pre-processing used by KMP is to calculate an array of the length of <strong>longest prefix suffix</strong> at each position of the pattern.</p>

<blockquote>
<p>LPS array: the longest suffix of substring <strong>P[0..i]</strong> which is also a prefix of <strong>P</strong>, where <strong>P</strong> denotes a pattern of length n and i the i-th position.</p>
</blockquote>

<p>As described in Dan&rsquo;s book, LPS array can be derived from Z algorithm, which I refer to as <strong>KMP-Z-based</strong>. However, the more commonly-known method to calculate this array is without Z algorithm (as in most textbooks), I&rsquo;ll call this method as <strong>KMP-vanilla</strong>. In this article, I&rsquo;ll use Z algorithm to illustrate KMP and Boyer-Moore methods, as in Dan&rsquo;s book.</p>

<p>The Z algorithm itself is a string matching method which also requires linear time to find all matches. It first calculates a Z array, which is the longest prefix of substring P[i..(n-1)] which is also a prefix of P. To find the matches using Z algorithm, you can use a trick that stitches the pattern to the start of the target string and the Z array for the new string would tell you where the pattern is found in the target string.</p>

<blockquote>
<p>Z array: the longest prefix of substring <strong>P[i..(n-1)]</strong> which is also a prefix of <strong>P</strong>.</p>
</blockquote>

<p>You can see the link between Z array and the longest prefix suffix (LPS) array, and how one can obtain the LPS array from Z array. For more details on implementation, I&rsquo;ve implemented a C++ version of both Z-based KMP and vanilla KMP at my <a href="https://github.com/dearxxj/computational_genomics/blob/master/exact-match/include/exactMatch.cpp">GitHub</a>.</p>

<p>Boyer-Moore algorithm is different from KMP such that it compares the pattern and target string from right to left. Other than that, it uses different preprocessing strategies, which are called the <strong>bad character rule</strong> and <strong>good suffix rule</strong> (For more details, please refer to <a href="http://www.langmead-lab.org/teaching-materials/">Ben&rsquo;s tutorial</a>). The bad character rule finds the nearest character to the mismatched one in the pattern that matches the mismatched character in the target string. For example, if the character &ldquo;x&rdquo; is mismatched with the &ldquo;a&rdquo; in the pattern, then the pattern should shift to the nearest &ldquo;x&rdquo; before &ldquo;a&rdquo; for comparison. We can see the bad character rule also utilizes the information from the target string to accelerate matching.</p>

<div class="highlight"><pre class="chroma"><code class="language-python" data-lang="python"><span class="ln">1</span> 
<span class="ln">2</span><span class="n">Bad</span> <span class="n">character</span> <span class="n">rule</span> <span class="n">finds</span> <span class="n">the</span> <span class="n">nearest</span> <span class="n">x</span> <span class="n">matches</span> <span class="n">the</span> <span class="n">target</span> <span class="n">string</span>
<span class="ln">3</span><span class="n">S</span><span class="p">:</span> <span class="o">-</span> <span class="o">-</span> <span class="o">-</span>  <span class="o">-</span>  <span class="o">-</span> <span class="n">b</span> <span class="n">b</span> <span class="p">(</span><span class="n">x</span><span class="p">)</span>
<span class="ln">4</span>                    <span class="o">|</span>
<span class="ln">5</span><span class="n">P</span><span class="p">:</span> <span class="n">b</span> <span class="n">b</span> <span class="o">-</span> <span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">-</span> <span class="n">b</span> <span class="n">b</span>  <span class="n">a</span></code></pre></div>

<p>The good suffix rule, on the other hand, looks for a substring (need not to a <em>prefix</em>) of <strong>P[0..(i-1)]</strong> that matches a suffix of the <strong>P[0..(i-1)]</strong>, where <strong>i</strong> is the mismatched position. The good suffix rule is, in some sense, similar to the LPS array which is a preprocessing of the pattern only. A stronger version of it also requires the character after the substring does not match <strong>P[i]</strong> since <strong>P[i]</strong> is a mismatch to the target string. For an C++ implementation of Boyer-Moore, please refer to my <a href="https://github.com/dearxxj/computational_genomics/blob/master/exact-match/include/exactMatch.cpp">GitHub</a>.</p>

<div class="highlight"><pre class="chroma"><code class="language-python" data-lang="python"><span class="ln">1</span> 
<span class="ln">2</span><span class="n">Good</span> <span class="n">suffix</span> <span class="n">rule</span> <span class="n">finds</span> <span class="n">the</span> <span class="n">nearest</span> <span class="n">substring</span> <span class="n">matches</span> <span class="n">the</span> <span class="n">suffix</span> <span class="n">before</span> <span class="n">the</span> <span class="n">mismatch</span>
<span class="ln">3</span><span class="n">S</span><span class="p">:</span>  <span class="o">-</span> <span class="o">-</span>  <span class="o">-</span> <span class="o">-</span> <span class="o">-</span>  <span class="n">b</span> <span class="n">b</span>  <span class="n">x</span>
<span class="ln">4</span>                <span class="o">|</span> <span class="o">|</span>  <span class="o">|</span> 
<span class="ln">5</span><span class="n">P</span><span class="p">:</span> <span class="p">(</span><span class="n">b</span> <span class="n">b</span><span class="p">)</span> <span class="o">-</span> <span class="n">x</span> <span class="o">-</span> <span class="p">(</span><span class="n">b</span> <span class="n">b</span><span class="p">)</span> <span class="n">a</span></code></pre></div>

<p>We can see that Boyer-Moore, while being more complicated than KMP, utilizes more information both from the target string and pattern. Given a long pattern, especially when there are repeated subpatterns (like a DNA sequence), Boyer-Moore should in theory out-perform KMP. Another question is whether KMP is indeed much inferior to Boyer-Moore in practice. I performed two tests for the three algorithms: 1) Boyer-Moore, 2) KMP Z-based, and 3) KMP vanilla.</p>

<p>To make a fair comparison of run time, I only calculated the run time for the actual matching step, but not the preprocessing step. I assume the preprocessing time is <strong>Boyer-Moore &gt; KMP Z-based &gt; KMP vanilla</strong>. However, in practice, preprocessing time should be considered.</p>

<h2 id="test">Test</h2>

<p>The first task is to search ~ 10000 most common English words (<a href="https://github.com/first20hours/google-10000-english">link</a>) in the Shakespeare complete works (<a href="https://ocw.mit.edu/ans7870/6/6.006/s08/lecturenotes/files/t8.shakespeare.txt">link</a>), which contains around 5.5 million characters. All English words are lower-case. The character table used in Boyer-Moore contains only 26 lower-case letters. The second task is to locating ~ 5000 randomly generated sequences (from <a href="http://www.bioinformatics.org/sms2/random_dna.html?">Bioinformatics.org</a>) in human chromosome 20. The character table used in Boyer-Moore is just &ldquo;acgt&rdquo;.</p>

<ul>
<li><strong>Task 1</strong>: Search ~ 10000 most common English words in Shakespeare complete works</li>
<li><strong>Task 2</strong>: Search ~ 5000 random DNA sequences of length varied from 4 to 100 basepair in human chromosome 20</li>
</ul>

<h2 id="result">Result</h2>

<p>In task 1, I plot the run time for three methods against the word length. We noticed that Boyer-Moore is significantly slower than KMP for shorter words (&lt; 5 chars). Boyer-Moore&rsquo;s run time is reversely related to the word length and starts to catch up with KMP at word length of 10, and even out-performs KMP for very long words (&gt; 16 chars). In practice, users may often want to search sentences or even paragraphs. In those cases, Boyer-Moore should be much faster than KMP. For shorter words, Boyer-Moore&rsquo;s performance is not that bad. <strong>So it&rsquo;s fair to say that KMP is slower than Boyer-Moore in practice</strong>.</p>

<p><img src="/img/run-time-for-words-v2.png" alt="Figure 1" /></p>

<p>It should be unsurprising that Boyer-Moore is faster than KMP for task 2. Because for DNA sequences which are made up of only four letters, one would expect many repeated subpatterns, where the good suffix rule shines. However, it is a bit surprising that Boyer-Moore is slower than vanilla KMP for all DNA sequences except for sequences of 51-bp long (might be due to a big skip in those sequences).</p>

<p><img src="/img/run-time-for-sequences-v2.png" alt="Figure 1" /></p>

<p>In conclusion, Boyer-Moore is slower than KMP for very short patterns, but shows increased performance as the pattern length increases. It&rsquo;s surprising to find that <strong>the vanilla KMP is not only simple but also very fast in both tasks</strong>. I&rsquo;m sure there is plenty of room for optimization of my C++ implementation of Boyer-Moore to further reduce the gap between Boyer-Moore and KMP at shorter patterns. I&rsquo;m also curious about the algorithm being used in the search function of Microsoft WORD, or any other softwares.</p>

<p><strong>Update (2019-4-7)</strong>: It seems that there are simpler (maybe faster) ways to implement Boyer-Moore.</p>

<ul>
<li>[1]. A <a href="http://www-igm.univ-mlv.fr/~lecroq/string/node14.html">great resource</a> for all kinds of string matching algorithms.</li>
<li>[2]. A C implementation of Boyer-Moore in <a href="https://github.com/samtools/htslib/blob/develop/kstring.c">HTSlib</a> based on notes in [1].</li>
</ul>


        
          <div class="blog-tags">
            
              <a href="https://dearxxj.github.io/tags/algorithm/">Algorithm</a>&nbsp;
            
              <a href="https://dearxxj.github.io/tags/bioinfo/">Bioinfo</a>&nbsp;
            
              <a href="https://dearxxj.github.io/tags/sequencing/">Sequencing</a>&nbsp;
            
          </div>
        

        
            <hr/>
            <section id="social-share">
              <div class="list-inline footer-links">
                

<div class="share-box" aria-hidden="true">
    <ul class="share">
      
      <li>
        <a href="//twitter.com/share?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f4%2f&amp;text=Exact%20string%20matching%20algorithms%3a%20Boyer-Moore%20and%20KMP&amp;via=" target="_blank" title="Share on Twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.facebook.com/sharer/sharer.php?u=https%3a%2f%2fdearxxj.github.io%2fpost%2f4%2f" target="_blank" title="Share on Facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//reddit.com/submit?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f4%2f&amp;title=Exact%20string%20matching%20algorithms%3a%20Boyer-Moore%20and%20KMP" target="_blank" title="Share on Reddit">
          <i class="fab fa-reddit"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.linkedin.com/shareArticle?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f4%2f&amp;title=Exact%20string%20matching%20algorithms%3a%20Boyer-Moore%20and%20KMP" target="_blank" title="Share on LinkedIn">
          <i class="fab fa-linkedin"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.stumbleupon.com/submit?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f4%2f&amp;title=Exact%20string%20matching%20algorithms%3a%20Boyer-Moore%20and%20KMP" target="_blank" title="Share on StumbleUpon">
          <i class="fab fa-stumbleupon"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.pinterest.com/pin/create/button/?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f4%2f&amp;description=Exact%20string%20matching%20algorithms%3a%20Boyer-Moore%20and%20KMP" target="_blank" title="Share on Pinterest">
          <i class="fab fa-pinterest"></i>
        </a>
      </li>
    </ul>
  </div>
  

              </div>
            </section>
        

        
          
            
          

          
                  <h4 class="see-also">See also</h4>
                  <ul>
                
                
                    <li><a href="/post/6/">Demystify Hi-C Data Normalization</a></li>
                
                    <li><a href="/post/5/">Learning HTSlib (1)</a></li>
                
                    <li><a href="/post/2/">How does number of threads affect mapping time in BWA MEM</a></li>
                
              </ul>

          
        
      </article>

      
        <ul class="pager blog-pager">
          
            <li class="previous">
              <a href="https://dearxxj.github.io/post/3/" data-toggle="tooltip" data-placement="top" title="Similarity measurement of two sets of TADs">&larr; Previous Post</a>
            </li>
          
          
            <li class="next">
              <a href="https://dearxxj.github.io/post/5/" data-toggle="tooltip" data-placement="top" title="Learning HTSlib (1)">Next Post &rarr;</a>
            </li>
          
        </ul>
      


      
        
          
          <div class="disqus-comments">
            <div id="disqus_thread"></div>
<script type="application/javascript">
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "fsong" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
          </div>
          
        
        
      

    </div>
  </div>
</div>

      
<footer>
  <div class="container">
    <div class="row">
      <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
        <ul class="list-inline text-center footer-links">
          
              <li>
                <a href="https://github.com/dearxxj" title="GitHub">
                  <span class="fa-stack fa-lg">
                    <i class="fas fa-circle fa-stack-2x"></i>
                    <i class="fab fa-github fa-stack-1x fa-inverse"></i>
                  </span>
                </a>
              </li>
          
          <li>
            <a href="" title="RSS">
              <span class="fa-stack fa-lg">
                <i class="fas fa-circle fa-stack-2x"></i>
                <i class="fas fa-rss fa-stack-1x fa-inverse"></i>
              </span>
            </a>
          </li>
          
        </ul>
        <p class="credits copyright text-muted">
          
            
              <a href="https://dearxxj.github.io">Fan</a>
            
          

          &nbsp;&bull;&nbsp;&copy;
          
            2019
          

          
            &nbsp;&bull;&nbsp;
            <a href="https://dearxxj.github.io">Fan&#39;s Blog</a>
          
        </p>
        
        <p class="credits theme-by text-muted">
          <a href="https://gohugo.io">Hugo v0.59.1</a> powered &nbsp;&bull;&nbsp; Theme <a href="https://github.com/halogenica/beautifulhugo">Beautiful Hugo</a> adapted from <a href="https://deanattali.com/beautiful-jekyll/">Beautiful Jekyll</a>
          
        </p>
      </div>
    </div>
  </div>
</footer><script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.js" integrity="sha384-K3vbOmF2BtaVai+Qk37uypf7VrgBubhQreNQe9aGsz9lB63dIFiQVlJbr92dw2Lx" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/contrib/auto-render.min.js" integrity="sha384-kmZOZB5ObwgQnS/DuDg6TScgOiWWBiVt0plIRkZCmE6rDZGrEOQeHM5PcHi+nyqe" crossorigin="anonymous"></script>
<script src="https://code.jquery.com/jquery-1.12.4.min.js" integrity="sha256-ZosEbRLbNQzLpnKIkEdrPv7lOy9C27hHQ+Xp8a4MxAQ=" crossorigin="anonymous"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<script src="https://dearxxj.github.io/js/main.js"></script>
<script src="https://dearxxj.github.io/js/highlight.min.js"></script>
<script> hljs.initHighlightingOnLoad(); </script>
<script> $(document).ready(function() {$("pre.chroma").css("padding","0");}); </script><script> renderMathInElement(document.body); </script><script src="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe.min.js" integrity="sha384-QELNnmcmU8IR9ZAykt67vGr9/rZJdHbiWi64V88fCPaOohUlHCqUD/unNN0BXSqy" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe-ui-default.min.js" integrity="sha384-m67o7SkQ1ALzKZIFh4CiTA8tmadaujiTa9Vu+nqPSwDOqHrDmxLezTdFln8077+q" crossorigin="anonymous"></script><script src="https://dearxxj.github.io/js/load-photoswipe.js"></script>









    
  </body>
</html>

