<!DOCTYPE html>
<html lang="en" itemscope itemtype="http://schema.org/WebPage">
  <head>
    

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0">

  <title>Learning HTSlib (1) - Fan&#39;s Blog</title>
  <meta name="description" content="In the past week, I&rsquo;ve been adapting my own HiC pipeline to the 4DN recommended pipeline. One critical step is to convert hundreds of read-paired BAM files generated by my old pipeline to pairs format. 4DN consortium provides a tool, bam2pairs, for this task. It&rsquo;s basically a Perl script that calls Samtools to read a BAM file and output targeted fields. Because the pair format suggests the columns to be sorted by chr1-chr2-pos1-pos2, bam2pairs then calls Linux &ldquo;sort&rdquo; to perform four times of sorting based on these four columns[1].">
  <meta name="author" content="Fan"/><script type="application/ld+json">
{
    "@context": "http://schema.org",
    "@type": "WebSite",
    "name": "Fan\x27s Blog",
    
    "url": "https:\/\/dearxxj.github.io"
}
</script><script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Organization",
  "name": "",
  "url": "https:\/\/dearxxj.github.io"
  
  
  
  
}
</script>
<script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [{
        "@type": "ListItem",
        "position": 1,
        "item": {
          "@id": "https:\/\/dearxxj.github.io",
          "name": "home"
        }
    },{
        "@type": "ListItem",
        "position": 3,
        "item": {
          "@id": "https:\/\/dearxxj.github.io\/post\/5\/",
          "name": "Learning h t slib (1)"
        }
    }]
}
</script><script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Article",
  "author": {
    "name" : "Fan"
  },
  "headline": "Learning HTSlib (1)",
  "description" : "In the past week, I\x26rsquo;ve been adapting my own HiC pipeline to the 4DN recommended pipeline. One critical step is to convert hundreds of read-paired BAM files generated by my old pipeline to pairs format. 4DN consortium provides a tool, bam2pairs, for this task. It\x26rsquo;s basically a Perl script that calls Samtools to read a BAM file and output targeted fields. Because the pair format suggests the columns to be sorted by chr1-chr2-pos1-pos2, bam2pairs then calls Linux \x26ldquo;sort\x26rdquo; to perform four times of sorting based on these four columns[1].",
  "inLanguage" : "en",
  "wordCount":  1219 ,
  "datePublished" : "2019-04-07T20:36:43",
  "dateModified" : "2019-04-07T20:36:43",
  "image" : "https:\/\/dearxxj.github.io\/img\/blog-log.png",
  "keywords" : [ "Bioinfo, C CPP, Sequencing" ],
  "mainEntityOfPage" : "https:\/\/dearxxj.github.io\/post\/5\/",
  "publisher" : {
    "@type": "Organization",
    "name" : "https:\/\/dearxxj.github.io",
    "logo" : {
        "@type" : "ImageObject",
        "url" : "https:\/\/dearxxj.github.io\/img\/blog-log.png",
        "height" :  60 ,
        "width" :  60
    }
  }
}
</script>

<meta property="og:title" content="Learning HTSlib (1)" />
<meta property="og:description" content="In the past week, I&rsquo;ve been adapting my own HiC pipeline to the 4DN recommended pipeline. One critical step is to convert hundreds of read-paired BAM files generated by my old pipeline to pairs format. 4DN consortium provides a tool, bam2pairs, for this task. It&rsquo;s basically a Perl script that calls Samtools to read a BAM file and output targeted fields. Because the pair format suggests the columns to be sorted by chr1-chr2-pos1-pos2, bam2pairs then calls Linux &ldquo;sort&rdquo; to perform four times of sorting based on these four columns[1].">
<meta property="og:image" content="https://dearxxj.github.io/img/blog-log.png" />
<meta property="og:url" content="https://dearxxj.github.io/post/5/" />
<meta property="og:type" content="website" />
<meta property="og:site_name" content="Fan&#39;s Blog" />

  <meta name="twitter:title" content="Learning HTSlib (1)" />
  <meta name="twitter:description" content="In the past week, I&rsquo;ve been adapting my own HiC pipeline to the 4DN recommended pipeline. One critical step is to convert hundreds of read-paired BAM files generated by my old pipeline to pairs â€¦">
  <meta name="twitter:image" content="https://dearxxj.github.io/img/blog-log.png" />
  <meta name="twitter:card" content="summary" />
  <link href='https://dearxxj.github.io/img/favicon.ico' rel='icon' type='image/x-icon'/>
  <meta name="generator" content="Hugo 0.59.1" />
  <link rel="alternate" href="https://dearxxj.github.io/index.xml" type="application/rss+xml" title="Fan&#39;s Blog"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.css" integrity="sha384-9eLZqc9ds8eNjO3TmqPeYcDj8n+Qfa4nuSiGYa6DjLNcv9BtN69ZIulL9+8CqC9Y" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.5.0/css/all.css" integrity="sha384-B4dIYHKNBt8Bc12p+WXckhzcICo0wtJAoU8YZTY5qE0Id1GSseTk6S+L3BlXeVIU" crossorigin="anonymous">
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><link rel="stylesheet" href="https://dearxxj.github.io/css/main.css" /><link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lora:400,700,400italic,700italic" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,600italic,700italic,800italic,400,300,600,700,800" />
  <link rel="stylesheet" href="https://dearxxj.github.io/css/highlight.min.css" /><link rel="stylesheet" href="https://dearxxj.github.io/css/codeblock.css" /><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe.min.css" integrity="sha384-h/L2W9KefUClHWaty3SLE5F/qvc4djlyR4qY3NUV5HGQBBW7stbcfff1+I/vmsHh" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/default-skin/default-skin.min.css" integrity="sha384-iD0dNku6PYSIQLyfTOpB06F2KCZJAKLOThS5HRe8b3ibhdEQ6eKsFf/EeFxdOt5R" crossorigin="anonymous">


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-104430016-1', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>


  </head>
  <body>
    <nav class="navbar navbar-default navbar-fixed-top navbar-custom">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#main-navbar">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="https://dearxxj.github.io">Fan&#39;s Blog</a>
    </div>

    <div class="collapse navbar-collapse" id="main-navbar">
      <ul class="nav navbar-nav navbar-right">
        
          
            <li>
              <a title="About" href="/page/about/">About</a>
            </li>
          
        
          
            <li>
              <a title="Tags" href="/tags">Tags</a>
            </li>
          
        

        

        
      </ul>
    </div>

    
      <div class="avatar-container">
        <div class="avatar-img-border">
          <a title="Fan&#39;s Blog" href="https://dearxxj.github.io">
            <img class="avatar-img" src="https://dearxxj.github.io/img/blog-log.png" alt="Fan&#39;s Blog" />
          </a>
        </div>
      </div>
    

  </div>
</nav>




    


<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

<div class="pswp__bg"></div>

<div class="pswp__scroll-wrap">
    
    <div class="pswp__container">
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
    </div>
    
    <div class="pswp__ui pswp__ui--hidden">
    <div class="pswp__top-bar">
      
      <div class="pswp__counter"></div>
      <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>
      <button class="pswp__button pswp__button--share" title="Share"></button>
      <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>
      <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>
      
      
      <div class="pswp__preloader">
        <div class="pswp__preloader__icn">
          <div class="pswp__preloader__cut">
            <div class="pswp__preloader__donut"></div>
          </div>
        </div>
      </div>
    </div>
    <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
      <div class="pswp__share-tooltip"></div>
    </div>
    <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
    </button>
    <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
    </button>
    <div class="pswp__caption">
      <div class="pswp__caption__center"></div>
    </div>
    </div>
    </div>
</div>


  
  
  






  

  <header class="header-section ">
    
    <div class="intro-header no-img">
      <div class="container">
        <div class="row">
          <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
            <div class="post-heading">
              
                <h1>Learning HTSlib (1)</h1>
              
              
              
              
                <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on April 7, 2019
  
  
    &nbsp;|&nbsp;<i class="fas fa-clock"></i>&nbsp;6&nbsp;minutes
  
  
    &nbsp;|&nbsp;<i class="fas fa-book"></i>&nbsp;1219&nbsp;words
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;Fan
    
  
  
</span>


              
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>


    
<div class="container" role="main">
  <div class="row">
    <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
      <article role="main" class="blog-post">
        

<p>In the past week, I&rsquo;ve been adapting my own HiC pipeline to the <a href="https://data.4dnucleome.org/help/analysis-and-visualization/hi_c-processing-pipeline">4DN recommended pipeline</a>. One critical step is to convert hundreds of read-paired BAM files generated by my old pipeline to <a href="https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md#pairs-file">pairs format</a>. 4DN consortium provides a tool, <a href="https://github.com/4dn-dcic/pairix/tree/master/util/bam2pairs">bam2pairs</a>, for this task. It&rsquo;s basically a Perl script that calls Samtools to read a BAM file and output targeted fields. Because the pair format suggests the columns to be sorted by <strong>chr1-chr2-pos1-pos2</strong>, bam2pairs then calls Linux &ldquo;sort&rdquo; to perform four times of sorting based on these four columns[1]. This step is actually very time-consuming. I tested it on a 200 million reads Hi-C bam file which took ~ 8 hours and 20 minutes to finish. This is an astounding task for my hundreds of BAM files, let alone many of these files contain billions of reads.</p>

<p>The reason that the pairs format is sorted this way is that some downstream tools (pairix/pairtools) requires the reads from the same block of the super matrix being placed together, such as reads from chr12-chr12, chr17-chr20, etc. Juicer tools (<a href="https://github.com/aidenlab/juicer/wiki/Pre">pre</a>) that generate .hic format from pairs format also requires input the pairs file to be sorted at least by <strong>chr1-chr2</strong>. However, <a href="https://cooler.readthedocs.io/en/latest/cli.html#cooler-cload-pairs">cooler</a> has no such requirements. Pairs files are no more than intermediate files for generating .hic or .cool binary files. What I need is just pair files sorted by <strong>chr1-chr2</strong>, this should save some time compared with four-column sorting. Now that the goal is clear, I can either modify the original bam2pairs script or write my own version. Since I haven&rsquo;t touched Perl for a long time, I think it&rsquo;s time to write it from scratch with either C (<a href="https://github.com/samtools/htslib">HTSlib</a>) or python (<a href="https://pysam.readthedocs.io/en/latest/api.html">pysam</a>). For efficiency consideration, I chose HTSlib for this task.</p>

<p>HTSlib is written by Heng Li, the author of Samtools and BWA. I wanted to share a little story here about him on how he influences the Bioinformatics field. The professor for my first-year Bioinformatics class once called him the &ldquo;Mapping God&rdquo;. Tools developed by Heng are fast and easy to use, like BWA/samtools/seqtk/bcftools/minimap etc. HTSlib is the underlying library for samtools/bcftools. One thing I don&rsquo;t like about HTSlib is that it does not have a good document. Maybe that&rsquo;s not a problem for pro developers. It gave me a hard time to learn HTSlib purely from the source code, but the process is lots of fun and I wanted to share my notes on HTSlib here.</p>

<p>HTSlib consists of a bunch of header files. Almost all SAM/BAM manipulations are written in the <a href="https://github.com/samtools/htslib/blob/develop/htslib/sam.h"><em>sam.h</em></a> file. If all you need is read and write sam files, this is probably the only header file that you need to include. It first defines a BAM header struct.</p>

<div class="highlight"><pre class="chroma"><code class="language-c" data-lang="c"><span class="ln"> 1</span><span class="k">typedef</span> <span class="k">struct</span> <span class="n">bam_hdr_t</span> <span class="p">{</span>
<span class="ln"> 2</span>    <span class="n">int32_t</span> <span class="n">n_targets</span><span class="p">,</span> <span class="n">ignore_sam_err</span><span class="p">;</span>  <span class="c1">// number of reference sequences
</span><span class="ln"> 3</span><span class="c1"></span>    <span class="n">uint32_t</span> <span class="n">l_text</span><span class="p">;</span>
<span class="ln"> 4</span>    <span class="n">uint32_t</span> <span class="o">*</span><span class="n">target_len</span><span class="p">;</span>
<span class="ln"> 5</span>    <span class="n">int8_t</span> <span class="o">*</span><span class="n">cigar_tab</span><span class="p">;</span>
<span class="ln"> 6</span>    <span class="kt">char</span> <span class="o">**</span><span class="n">target_name</span><span class="p">;</span>  <span class="c1">// names of the reference sequences
</span><span class="ln"> 7</span><span class="c1"></span>    <span class="kt">char</span> <span class="o">*</span><span class="n">text</span><span class="p">;</span>
<span class="ln"> 8</span>    <span class="kt">void</span> <span class="o">*</span><span class="n">sdict</span><span class="p">;</span>      <span class="c1">// header dictionary
</span><span class="ln"> 9</span><span class="c1"></span>    <span class="n">uint32_t</span> <span class="n">ref_count</span><span class="p">;</span>
<span class="ln">10</span><span class="p">}</span> <span class="n">bam_hdr_t</span><span class="p">;</span></code></pre></div>

<p><em>target_name</em> is an array of references names such as chr1, chr2, &hellip;, etc. It is useful when one wants to retrieve chromosome names from sam records which only stores the name indexes to save space. sdict is a string to integer hash table which is used in header parsing. Followed by the header struct are CIGAR-related macros.</p>

<div class="highlight"><pre class="chroma"><code class="language-c" data-lang="c"><span class="ln"> 1</span><span class="cp">#define BAM_FPAIRED        1
</span><span class="ln"> 2</span><span class="cp"></span><span class="cm">/*! @abstract the read is mapped in a proper pair */</span>
<span class="ln"> 3</span><span class="cp">#define BAM_FPROPER_PAIR   2
</span><span class="ln"> 4</span><span class="cp"></span><span class="cm">/*! @abstract the read itself is unmapped; conflictive with BAM_FPROPER_PAIR */</span>
<span class="ln"> 5</span><span class="cp">#define BAM_FUNMAP         4
</span><span class="ln"> 6</span><span class="cp"></span><span class="cm">/*! @abstract the mate is unmapped */</span>
<span class="ln"> 7</span><span class="cp">#define BAM_FMUNMAP        8
</span><span class="ln"> 8</span><span class="cp"></span><span class="cm">/*! @abstract the read is mapped to the reverse strand */</span>
<span class="ln"> 9</span><span class="cp">#define BAM_FREVERSE      16
</span><span class="ln">10</span><span class="cp"></span><span class="cm">/*! @abstract the mate is mapped to the reverse strand */</span>
<span class="ln">11</span><span class="cp">#define BAM_FMREVERSE     32</span></code></pre></div>

<p>I found BAM_FUNMAP and BAM_FMUNMAP are useful for basic read filtering, which is equivalent to &ldquo;samtools view -F12&rdquo;. BAM_FREVERSE and BAM_FMREVERSE can be used to distinguish reads mapped to two strands. The way to do these filtering in C is simply &ldquo;flag &amp; BAM_XXXXX&rdquo;.</p>

<p>Next, structs for one alignment and alignment core information are defined.</p>

<div class="highlight"><pre class="chroma"><code class="language-c" data-lang="c"><span class="ln"> 1</span><span class="k">typedef</span> <span class="k">struct</span> <span class="p">{</span>
<span class="ln"> 2</span>    <span class="n">int32_t</span> <span class="n">tid</span><span class="p">;</span>    <span class="c1">// chromosome ID, defined by bam_hdr_t
</span><span class="ln"> 3</span><span class="c1"></span>    <span class="n">int32_t</span> <span class="n">pos</span><span class="p">;</span>    <span class="c1">// 0-based leftmost coordinate
</span><span class="ln"> 4</span><span class="c1"></span>    <span class="n">uint16_t</span> <span class="n">bin</span><span class="p">;</span>
<span class="ln"> 5</span>    <span class="n">uint8_t</span> <span class="n">qual</span><span class="p">;</span>
<span class="ln"> 6</span>    <span class="n">uint8_t</span> <span class="n">l_qname</span><span class="p">;</span>
<span class="ln"> 7</span>    <span class="n">uint16_t</span> <span class="n">flag</span><span class="p">;</span>
<span class="ln"> 8</span>    <span class="n">uint8_t</span> <span class="n">unused1</span><span class="p">;</span>
<span class="ln"> 9</span>    <span class="n">uint8_t</span> <span class="n">l_extranul</span><span class="p">;</span>
<span class="ln">10</span>    <span class="n">uint32_t</span> <span class="n">n_cigar</span><span class="p">;</span>
<span class="ln">11</span>    <span class="n">int32_t</span> <span class="n">l_qseq</span><span class="p">;</span>
<span class="ln">12</span>    <span class="n">int32_t</span> <span class="n">mtid</span><span class="p">;</span>   <span class="c1">// chromosome ID of next/mate read in template, defined by bam_hdr_t
</span><span class="ln">13</span><span class="c1"></span>    <span class="n">int32_t</span> <span class="n">mpos</span><span class="p">;</span>
<span class="ln">14</span>    <span class="n">int32_t</span> <span class="n">isize</span><span class="p">;</span>
<span class="ln">15</span><span class="p">}</span> <span class="n">bam1_core_t</span><span class="p">;</span>
<span class="ln">16</span>
<span class="ln">17</span><span class="k">typedef</span> <span class="k">struct</span> <span class="p">{</span>
<span class="ln">18</span>    <span class="n">bam1_core_t</span> <span class="n">core</span><span class="p">;</span>  <span class="c1">// bam1_core_t
</span><span class="ln">19</span><span class="c1"></span>    <span class="kt">int</span> <span class="n">l_data</span><span class="p">;</span>
<span class="ln">20</span>    <span class="n">uint32_t</span> <span class="n">m_data</span><span class="p">;</span>
<span class="ln">21</span>    <span class="n">uint8_t</span> <span class="o">*</span><span class="n">data</span><span class="p">;</span> <span class="c1">// all variable-length data, concatenated; structure: qname-cigar-seq-qual-aux
</span><span class="ln">22</span><span class="c1"></span><span class="cp">#ifndef BAM_NO_ID
</span><span class="ln">23</span><span class="cp"></span>    <span class="n">uint64_t</span> <span class="n">id</span><span class="p">;</span>
<span class="ln">24</span><span class="cp">#endif
</span><span class="ln">25</span><span class="cp"></span><span class="p">}</span> <span class="n">bam1_t</span><span class="p">;</span></code></pre></div>

<p>All fields in the bam1_core_t are pretty self-explained. <em>tid/mtid</em> are indexes to the chromosome names as mentioned above (e.g. <em>bam_hdr-&gt;target_names[tid]</em>). Note that <em>pos</em> here is 0-based while SAM format is 1-based, therefore (<em>pos</em> + 1) is need to generate SAM-like output. <em>bam1_t</em> is defined in a way similar to <em>kstring_t</em> struct. It can be accessed by some useful APIs defined in the following lines.</p>

<div class="highlight"><pre class="chroma"><code class="language-c" data-lang="c"><span class="ln">1</span><span class="cp">#define bam_is_rev(b) (((b)-&gt;core.flag&amp;BAM_FREVERSE) != 0)  </span><span class="c1">// whether the query is on the reverse strand
</span><span class="ln">2</span><span class="c1"></span><span class="cp">#define bam_is_mrev(b) (((b)-&gt;core.flag&amp;BAM_FMREVERSE) != 0) </span><span class="c1">// whether the mate query is on the reverse strand
</span><span class="ln">3</span><span class="c1"></span><span class="cp">#define bam_get_qname(b) ((char*)(b)-&gt;data)                  </span><span class="c1">// Get the name of the query
</span><span class="ln">4</span><span class="c1"></span><span class="cp">#define bam_get_cigar(b)                                     </span><span class="c1">// Get the CIGAR array
</span><span class="ln">5</span><span class="c1"></span><span class="cp">#define bam_get_seq(b)                                       </span><span class="c1">// Get query sequence
</span><span class="ln">6</span><span class="c1"></span><span class="cp">#define bam_seqi(s, i) ((s)[(i)&gt;&gt;1] &gt;&gt; ((~(i)&amp;1)&lt;&lt;2) &amp; 0xf)  // Get a base on read</span></code></pre></div>

<p>There are at least two ways to read a BAM file: <strong>1)</strong> read from the beginning; <strong>2)</strong> read from a given chromosomal position by loading the BAM index.</p>

<div class="highlight"><pre class="chroma"><code class="language-c" data-lang="c"><span class="ln"> 1</span><span class="c1">////////// either way, initialization of an alignment is needed
</span><span class="ln"> 2</span><span class="c1"></span><span class="n">b</span> <span class="o">=</span> <span class="n">bam_init1</span><span class="p">();</span>
<span class="ln"> 3</span><span class="n">in</span> <span class="o">=</span> <span class="n">sam_open</span><span class="p">(</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s">&#34;r&#34;</span><span class="p">);</span>
<span class="ln"> 4</span><span class="k">if</span> <span class="p">(</span><span class="n">in</span> <span class="o">==</span> <span class="nb">NULL</span><span class="p">)</span> <span class="k">return</span> <span class="o">-</span><span class="mi">1</span><span class="p">;</span>
<span class="ln"> 5</span><span class="n">header</span> <span class="o">=</span> <span class="n">sam_hdr_read</span><span class="p">(</span><span class="n">in</span><span class="p">);</span>
<span class="ln"> 6</span><span class="k">if</span> <span class="p">(</span><span class="n">header</span> <span class="o">=</span> <span class="nb">NULL</span><span class="p">)</span> <span class="k">return</span> <span class="o">-</span><span class="mi">1</span><span class="p">;</span>
<span class="ln"> 7</span>
<span class="ln"> 8</span><span class="c1">////////// 1. Read from the beginning
</span><span class="ln"> 9</span><span class="c1"></span><span class="k">while</span> <span class="p">(</span><span class="n">sam_read1</span><span class="p">(</span><span class="n">in</span><span class="p">,</span> <span class="n">header</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
<span class="ln">10</span>    <span class="n">c</span> <span class="o">=</span> <span class="o">&amp;</span><span class="n">b</span><span class="o">-&gt;</span><span class="n">core</span><span class="p">;</span>
<span class="ln">11</span>    <span class="n">chr</span> <span class="o">=</span> <span class="n">header</span><span class="o">-&gt;</span><span class="n">target_name</span><span class="p">[</span><span class="n">c</span><span class="o">-&gt;</span><span class="n">tid</span><span class="p">];</span>
<span class="ln">12</span>    <span class="n">mchr</span> <span class="o">=</span> <span class="n">header</span><span class="o">-&gt;</span><span class="n">target_name</span><span class="p">[</span><span class="n">c</span><span class="o">-&gt;</span><span class="n">mtid</span><span class="p">];</span>
<span class="ln">13</span>    <span class="n">pos</span> <span class="o">=</span> <span class="n">c</span><span class="o">-&gt;</span><span class="n">pos</span> <span class="o">+</span> <span class="mi">1</span><span class="p">;</span> <span class="c1">// sam format is 1-based
</span><span class="ln">14</span><span class="c1"></span>    <span class="n">mpos</span> <span class="o">=</span> <span class="n">c</span><span class="o">-&gt;</span><span class="n">mpos</span> <span class="o">+</span> <span class="mi">1</span><span class="p">;</span>
<span class="ln">15</span>    <span class="n">fprintf</span><span class="p">(</span><span class="n">stdout</span><span class="p">,</span> <span class="s">&#34;%s</span><span class="se">\t</span><span class="s">%d</span><span class="se">\t</span><span class="s">%s</span><span class="se">\t</span><span class="s">%d</span><span class="se">\n</span><span class="s">&#34;</span><span class="p">,</span> <span class="n">chr</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">mchr</span><span class="p">,</span> <span class="n">mpos</span><span class="p">);</span>
<span class="ln">16</span><span class="p">}</span>
<span class="ln">17</span>
<span class="ln">18</span><span class="c1">////////// 2. Read from a given region
</span><span class="ln">19</span><span class="c1"></span><span class="n">hts_itr_t</span> <span class="o">*</span><span class="n">iter</span><span class="p">;</span>    <span class="c1">// initiate an iterator
</span><span class="ln">20</span><span class="c1"></span><span class="n">hts_idx_t</span> <span class="o">*</span><span class="n">idx</span><span class="p">;</span>  <span class="c1">// initiate an BAM index
</span><span class="ln">21</span><span class="c1"></span><span class="n">idx</span> <span class="o">=</span> <span class="n">sam_index_load</span><span class="p">(</span><span class="n">in</span><span class="p">,</span> <span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">]);</span>   <span class="c1">// second parameter is same as BAM file path
</span><span class="ln">22</span><span class="c1"></span><span class="k">if</span> <span class="p">(</span><span class="n">idx</span> <span class="o">==</span> <span class="nb">NULL</span><span class="p">)</span> <span class="k">return</span> <span class="o">-</span><span class="mi">1</span><span class="p">;</span>
<span class="ln">23</span><span class="n">iter</span>  <span class="o">=</span> <span class="n">sam_itr_querys</span><span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">header</span><span class="p">,</span> <span class="n">argv</span><span class="p">[</span><span class="mi">2</span><span class="p">]);</span>    <span class="c1">//  third parameter is like chr1:1000-2000, locate to the start of the region
</span><span class="ln">24</span><span class="c1"></span><span class="k">while</span> <span class="p">(</span> <span class="n">sam_itr_next</span><span class="p">(</span><span class="n">in</span><span class="p">,</span> <span class="n">iter</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
<span class="ln">25</span>    <span class="p">...</span>
<span class="ln">26</span><span class="p">}</span>
<span class="ln">27</span>
<span class="ln">28</span><span class="c1">////////// Last, don&#39;t forget to destroy all pointers
</span><span class="ln">29</span><span class="c1"></span><span class="n">hts_itr_destroy</span><span class="p">(</span><span class="n">iter</span><span class="p">);</span>
<span class="ln">30</span><span class="n">bam_destroy1</span><span class="p">(</span><span class="n">b</span><span class="p">);</span>
<span class="ln">31</span><span class="n">bam_hdr_destroy</span><span class="p">(</span><span class="n">header</span><span class="p">);</span>
<span class="ln">32</span><span class="n">sam_close</span><span class="p">(</span><span class="n">in</span><span class="p">);</span></code></pre></div>

<p><br></p>

<h5 id="compile-a-htslib-program">Compile a HTSlib program</h5>

<p><div class="highlight"><pre class="chroma"><code class="language-shell" data-lang="shell"><span class="ln">1</span><span class="c1">#  1. compile HTSlib and create a static library and header files</span>
<span class="ln">2</span>git clone ....<span class="p">;</span> ...<span class="p">;</span> make <span class="o">&amp;&amp;</span> make install
<span class="ln">3</span>
<span class="ln">4</span><span class="c1"># 2. link to the HTSlib library and other necessary libraries</span>
<span class="ln">5</span>gcc -O2 -Wall main.c htslib-build/lib/libhts.a -lz -lbz2 -llzma -lcurl -o bam2pairs</code></pre></div>
<br></p>

<p><br>
[1]. (Update 191103). GNU sort actually can be pretty fast if used properly (<a href="https://unix.stackexchange.com/questions/120096/how-to-sort-big-files">ref</a>) since it is already optimized for large files. To do this, one needs to set the &ndash;parallel parameter to the number of cores that sort will use. Algorithmly, GNU sort implements external sorting which first split the input into smaller portions and fit that in the memory, then perform sorting on each portion and lastly merge the results.</p>


        
          <div class="blog-tags">
            
              <a href="https://dearxxj.github.io/tags/bioinfo/">Bioinfo</a>&nbsp;
            
              <a href="https://dearxxj.github.io/tags/c-cpp/">C CPP</a>&nbsp;
            
              <a href="https://dearxxj.github.io/tags/sequencing/">Sequencing</a>&nbsp;
            
          </div>
        

        
            <hr/>
            <section id="social-share">
              <div class="list-inline footer-links">
                

<div class="share-box" aria-hidden="true">
    <ul class="share">
      
      <li>
        <a href="//twitter.com/share?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f5%2f&amp;text=Learning%20HTSlib%20%281%29&amp;via=" target="_blank" title="Share on Twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.facebook.com/sharer/sharer.php?u=https%3a%2f%2fdearxxj.github.io%2fpost%2f5%2f" target="_blank" title="Share on Facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//reddit.com/submit?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f5%2f&amp;title=Learning%20HTSlib%20%281%29" target="_blank" title="Share on Reddit">
          <i class="fab fa-reddit"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.linkedin.com/shareArticle?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f5%2f&amp;title=Learning%20HTSlib%20%281%29" target="_blank" title="Share on LinkedIn">
          <i class="fab fa-linkedin"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.stumbleupon.com/submit?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f5%2f&amp;title=Learning%20HTSlib%20%281%29" target="_blank" title="Share on StumbleUpon">
          <i class="fab fa-stumbleupon"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.pinterest.com/pin/create/button/?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f5%2f&amp;description=Learning%20HTSlib%20%281%29" target="_blank" title="Share on Pinterest">
          <i class="fab fa-pinterest"></i>
        </a>
      </li>
    </ul>
  </div>
  

              </div>
            </section>
        

        
          
            
          

          
                  <h4 class="see-also">See also</h4>
                  <ul>
                
                
                    <li><a href="/post/4/">Exact string matching algorithms: Boyer-Moore and KMP</a></li>
                
                    <li><a href="/post/2/">How does number of threads affect mapping time in BWA MEM</a></li>
                
              </ul>

          
        
      </article>

      
        <ul class="pager blog-pager">
          
            <li class="previous">
              <a href="https://dearxxj.github.io/post/4/" data-toggle="tooltip" data-placement="top" title="Exact string matching algorithms: Boyer-Moore and KMP">&larr; Previous Post</a>
            </li>
          
          
        </ul>
      


      
        
          
          <div class="disqus-comments">
            <div id="disqus_thread"></div>
<script type="application/javascript">
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "fsong" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
          </div>
          
        
        
      

    </div>
  </div>
</div>

      
<footer>
  <div class="container">
    <div class="row">
      <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
        <ul class="list-inline text-center footer-links">
          
              <li>
                <a href="https://github.com/dearxxj" title="GitHub">
                  <span class="fa-stack fa-lg">
                    <i class="fas fa-circle fa-stack-2x"></i>
                    <i class="fab fa-github fa-stack-1x fa-inverse"></i>
                  </span>
                </a>
              </li>
          
          <li>
            <a href="" title="RSS">
              <span class="fa-stack fa-lg">
                <i class="fas fa-circle fa-stack-2x"></i>
                <i class="fas fa-rss fa-stack-1x fa-inverse"></i>
              </span>
            </a>
          </li>
          
        </ul>
        <p class="credits copyright text-muted">
          
            
              <a href="https://dearxxj.github.io">Fan</a>
            
          

          &nbsp;&bull;&nbsp;&copy;
          
            2019
          

          
            &nbsp;&bull;&nbsp;
            <a href="https://dearxxj.github.io">Fan&#39;s Blog</a>
          
        </p>
        
        <p class="credits theme-by text-muted">
          <a href="https://gohugo.io">Hugo v0.59.1</a> powered &nbsp;&bull;&nbsp; Theme <a href="https://github.com/halogenica/beautifulhugo">Beautiful Hugo</a> adapted from <a href="https://deanattali.com/beautiful-jekyll/">Beautiful Jekyll</a>
          
        </p>
      </div>
    </div>
  </div>
</footer><script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.js" integrity="sha384-K3vbOmF2BtaVai+Qk37uypf7VrgBubhQreNQe9aGsz9lB63dIFiQVlJbr92dw2Lx" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/contrib/auto-render.min.js" integrity="sha384-kmZOZB5ObwgQnS/DuDg6TScgOiWWBiVt0plIRkZCmE6rDZGrEOQeHM5PcHi+nyqe" crossorigin="anonymous"></script>
<script src="https://code.jquery.com/jquery-1.12.4.min.js" integrity="sha256-ZosEbRLbNQzLpnKIkEdrPv7lOy9C27hHQ+Xp8a4MxAQ=" crossorigin="anonymous"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<script src="https://dearxxj.github.io/js/main.js"></script>
<script src="https://dearxxj.github.io/js/highlight.min.js"></script>
<script> hljs.initHighlightingOnLoad(); </script>
<script> $(document).ready(function() {$("pre.chroma").css("padding","0");}); </script><script> renderMathInElement(document.body); </script><script src="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe.min.js" integrity="sha384-QELNnmcmU8IR9ZAykt67vGr9/rZJdHbiWi64V88fCPaOohUlHCqUD/unNN0BXSqy" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe-ui-default.min.js" integrity="sha384-m67o7SkQ1ALzKZIFh4CiTA8tmadaujiTa9Vu+nqPSwDOqHrDmxLezTdFln8077+q" crossorigin="anonymous"></script><script src="https://dearxxj.github.io/js/load-photoswipe.js"></script>









    
  </body>
</html>

