<!DOCTYPE html>
<html lang="en" itemscope itemtype="http://schema.org/WebPage">
  <head>
    

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0">

  <title>Demystify Hi-C Data Normalization - Fan&#39;s Blog</title>
  <meta name="description" content="Hi-C is a sequencing-based method for profiling the genome-wide chromatin contacts. It has been widely used in studying various biological questions such as gene regulation, chromatin structures, genome assembly, etc. The Hi-C experiments involves a series of biochemistical reactions that may introduce noises to the output. Subsequent data analysis such as read mapping also give rise to noises that affect the interpretation of the final output: a contact matrix, where each element in the matrix denotes the contact strength between any two regions of genome.">
  <meta name="author" content="Fan"/><script type="application/ld+json">
{
    "@context": "http://schema.org",
    "@type": "WebSite",
    "name": "Fan\x27s Blog",
    
    "url": "https:\/\/dearxxj.github.io"
}
</script><script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Organization",
  "name": "",
  "url": "https:\/\/dearxxj.github.io"
  
  
  
  
}
</script>
<script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [{
        "@type": "ListItem",
        "position": 1,
        "item": {
          "@id": "https:\/\/dearxxj.github.io",
          "name": "home"
        }
    },{
        "@type": "ListItem",
        "position": 3,
        "item": {
          "@id": "https:\/\/dearxxj.github.io\/post\/6\/",
          "name": "Demystify hi c data normalization"
        }
    }]
}
</script><script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Article",
  "author": {
    "name" : "Fan"
  },
  "headline": "Demystify Hi-C Data Normalization",
  "description" : "Hi-C is a sequencing-based method for profiling the genome-wide chromatin contacts. It has been widely used in studying various biological questions such as gene regulation, chromatin structures, genome assembly, etc. The Hi-C experiments involves a series of biochemistical reactions that may introduce noises to the output. Subsequent data analysis such as read mapping also give rise to noises that affect the interpretation of the final output: a contact matrix, where each element in the matrix denotes the contact strength between any two regions of genome.",
  "inLanguage" : "en",
  "wordCount":  1402 ,
  "datePublished" : "2019-11-03T20:59:30",
  "dateModified" : "2019-11-03T20:59:30",
  "image" : "https:\/\/dearxxj.github.io\/img\/blog-log.png",
  "keywords" : [ "Bioinfo, Hi-C" ],
  "mainEntityOfPage" : "https:\/\/dearxxj.github.io\/post\/6\/",
  "publisher" : {
    "@type": "Organization",
    "name" : "https:\/\/dearxxj.github.io",
    "logo" : {
        "@type" : "ImageObject",
        "url" : "https:\/\/dearxxj.github.io\/img\/blog-log.png",
        "height" :  60 ,
        "width" :  60
    }
  }
}
</script>

<meta property="og:title" content="Demystify Hi-C Data Normalization" />
<meta property="og:description" content="Hi-C is a sequencing-based method for profiling the genome-wide chromatin contacts. It has been widely used in studying various biological questions such as gene regulation, chromatin structures, genome assembly, etc. The Hi-C experiments involves a series of biochemistical reactions that may introduce noises to the output. Subsequent data analysis such as read mapping also give rise to noises that affect the interpretation of the final output: a contact matrix, where each element in the matrix denotes the contact strength between any two regions of genome.">
<meta property="og:image" content="https://dearxxj.github.io/img/blog-log.png" />
<meta property="og:url" content="https://dearxxj.github.io/post/6/" />
<meta property="og:type" content="website" />
<meta property="og:site_name" content="Fan&#39;s Blog" />

  <meta name="twitter:title" content="Demystify Hi-C Data Normalization" />
  <meta name="twitter:description" content="Hi-C is a sequencing-based method for profiling the genome-wide chromatin contacts. It has been widely used in studying various biological questions such as gene regulation, chromatin structures, â€¦">
  <meta name="twitter:image" content="https://dearxxj.github.io/img/blog-log.png" />
  <meta name="twitter:card" content="summary" />
  <link href='https://dearxxj.github.io/img/favicon.ico' rel='icon' type='image/x-icon'/>
  <meta name="generator" content="Hugo 0.59.1" />
  <link rel="alternate" href="https://dearxxj.github.io/index.xml" type="application/rss+xml" title="Fan&#39;s Blog"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.css" integrity="sha384-9eLZqc9ds8eNjO3TmqPeYcDj8n+Qfa4nuSiGYa6DjLNcv9BtN69ZIulL9+8CqC9Y" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.5.0/css/all.css" integrity="sha384-B4dIYHKNBt8Bc12p+WXckhzcICo0wtJAoU8YZTY5qE0Id1GSseTk6S+L3BlXeVIU" crossorigin="anonymous">
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><link rel="stylesheet" href="https://dearxxj.github.io/css/main.css" /><link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lora:400,700,400italic,700italic" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,600italic,700italic,800italic,400,300,600,700,800" />
  <link rel="stylesheet" href="https://dearxxj.github.io/css/highlight.min.css" /><link rel="stylesheet" href="https://dearxxj.github.io/css/codeblock.css" /><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe.min.css" integrity="sha384-h/L2W9KefUClHWaty3SLE5F/qvc4djlyR4qY3NUV5HGQBBW7stbcfff1+I/vmsHh" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/default-skin/default-skin.min.css" integrity="sha384-iD0dNku6PYSIQLyfTOpB06F2KCZJAKLOThS5HRe8b3ibhdEQ6eKsFf/EeFxdOt5R" crossorigin="anonymous">


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-104430016-1', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>


  </head>
  <body>
    <nav class="navbar navbar-default navbar-fixed-top navbar-custom">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#main-navbar">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="https://dearxxj.github.io">Fan&#39;s Blog</a>
    </div>

    <div class="collapse navbar-collapse" id="main-navbar">
      <ul class="nav navbar-nav navbar-right">
        
          
            <li>
              <a title="About" href="/page/about/">About</a>
            </li>
          
        
          
            <li>
              <a title="Tags" href="/tags">Tags</a>
            </li>
          
        

        

        
      </ul>
    </div>

    
      <div class="avatar-container">
        <div class="avatar-img-border">
          <a title="Fan&#39;s Blog" href="https://dearxxj.github.io">
            <img class="avatar-img" src="https://dearxxj.github.io/img/blog-log.png" alt="Fan&#39;s Blog" />
          </a>
        </div>
      </div>
    

  </div>
</nav>




    


<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

<div class="pswp__bg"></div>

<div class="pswp__scroll-wrap">
    
    <div class="pswp__container">
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
    </div>
    
    <div class="pswp__ui pswp__ui--hidden">
    <div class="pswp__top-bar">
      
      <div class="pswp__counter"></div>
      <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>
      <button class="pswp__button pswp__button--share" title="Share"></button>
      <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>
      <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>
      
      
      <div class="pswp__preloader">
        <div class="pswp__preloader__icn">
          <div class="pswp__preloader__cut">
            <div class="pswp__preloader__donut"></div>
          </div>
        </div>
      </div>
    </div>
    <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
      <div class="pswp__share-tooltip"></div>
    </div>
    <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
    </button>
    <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
    </button>
    <div class="pswp__caption">
      <div class="pswp__caption__center"></div>
    </div>
    </div>
    </div>
</div>


  
  
  






  

  <header class="header-section ">
    
    <div class="intro-header no-img">
      <div class="container">
        <div class="row">
          <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
            <div class="post-heading">
              
                <h1>Demystify Hi-C Data Normalization</h1>
              
              
              
              
                <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on November 3, 2019
  
  
    &nbsp;|&nbsp;<i class="fas fa-clock"></i>&nbsp;7&nbsp;minutes
  
  
    &nbsp;|&nbsp;<i class="fas fa-book"></i>&nbsp;1402&nbsp;words
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;Fan
    
  
  
</span>


              
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>


    
<div class="container" role="main">
  <div class="row">
    <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
      <article role="main" class="blog-post">
        

<p>Hi-C is a sequencing-based method for profiling the genome-wide chromatin contacts. It has been widely used in studying various biological questions such as gene regulation, chromatin structures, genome assembly, etc. The Hi-C experiments involves a series of biochemistical reactions that may introduce noises to the output. Subsequent data analysis such as read mapping also give rise to noises that affect the interpretation of the final output: <strong>a contact matrix, where each element in the matrix denotes the contact strength between any two regions of genome</strong>. Therefore, a critical step in Hi-C data analysis is to remove such noises, a step also called Hi-C data normalization.</p>

<p>Early days of normalization methods for Hi-C data focused on explicit factors that cause noises. Factors like cutting enzyme sites, genome mappability, GC contents make the sequencing reads unevnely distributed across the genome, thus introduce biases when calculating pairwise contacts. In light of these ideas, Imakaev et al and others propose a method cappable of handling all sources of noises &ldquo;implicitly&rdquo; [1]. The idea behind is actually very simple: because Hi-C is an unbiased experiments in theory, <strong>the visibility across all genomic regions should be equal</strong>. Another assumption is that all biases of a Hi-C experiments are one-dimensional and factorizable. Based on these assumption, a natural solution is to factorize the raw contact matrix into a multiplication of two 1D biases and a normalized matrix whose rows and columns summed up to a same value.</p>

<p>The method proposed by Imakaev is also known as matrix balancing in matrix theory. Matrix balancing is an old mathematical question dated back to one century ago. Many matrix balancing methods have been developed including vanilla coverage (<strong>VC</strong>), Sinkhorn &amp; Knopp (<strong>SK</strong>), and Knight &amp; Ruiz methods (<strong>KR</strong>). VC is done by dividing each element of the matrix by its row sum and column sum to remove different sequencing coverage of each loci. VC can be considered as a single iteration of the SK method. In SK, the VC process is repeatly performed until all the rows and cols sum to the same value. S&amp;K proves that this process will converge. Imakaev et al uses the standard SK method.</p>

<p>Rao et al reviewed all matrix balancing method ands introduced the KR method to Hi-C data [2]. Based on the original paper by K&amp;R, KR method is several magnitute faster than SP, which makes it suitable for balancing high-resolution matrices. In practise, ICE impelementation of SP is also very fast even at 10kb resolution. In my experience, the speed difference between ICE and KR is neglectable. However, KR has a drawback that the KR process may not converge when the matrix is too sparse (as pointed out by Rao et al. 2014). This happened a couple of times in my research when I used juicertools to generate KR normalized matrix on a low sequencing dataset and got an empty matrix.</p>

<p>The algorithms for matrix balancing is actually not difficult, at least not as the mathematical theories behind it. So, how do we calculate a balanced matrix for a Hi-C contact matrix? Below I implemented VC and SP method in a short python class. This implementation is fast for small matrices.</p>

<div class="highlight"><pre class="chroma"><code class="language-python" data-lang="python"><span class="ln"> 1</span><span class="k">class</span> <span class="nc">HiCNorm</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="ln"> 2</span>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">matrix</span><span class="p">):</span>
<span class="ln"> 3</span>        <span class="bp">self</span><span class="o">.</span><span class="n">bias</span> <span class="o">=</span> <span class="bp">None</span>
<span class="ln"> 4</span>        <span class="bp">self</span><span class="o">.</span><span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix</span>
<span class="ln"> 5</span>        <span class="bp">self</span><span class="o">.</span><span class="n">norm_matrix</span> <span class="o">=</span> <span class="bp">None</span>
<span class="ln"> 6</span>        <span class="bp">self</span><span class="o">.</span><span class="n">_bias_var</span> <span class="o">=</span> <span class="bp">None</span>
<span class="ln"> 7</span>
<span class="ln"> 8</span>    <span class="k">def</span> <span class="nf">iterative_correction</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">max_iter</span><span class="o">=</span><span class="mi">50</span><span class="p">):</span>
<span class="ln"> 9</span>        <span class="n">mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">matrix</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>  <span class="c1">## use float to avoid int overflow</span>
<span class="ln">10</span>        <span class="c1"># remove low count rows for normalization</span>
<span class="ln">11</span>        <span class="n">row_sum</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="ln">12</span>        <span class="n">low_count</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">row_sum</span><span class="p">,</span> <span class="mf">0.15</span><span class="p">)</span>  <span class="c1"># 0.15 gives best correlation between KR/SP/VC</span>
<span class="ln">13</span>        <span class="n">mask_row</span> <span class="o">=</span> <span class="n">row_sum</span> <span class="o">&lt;</span> <span class="n">low_count</span>
<span class="ln">14</span>        <span class="n">mat</span><span class="p">[</span><span class="n">mask_row</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="mi">0</span>
<span class="ln">15</span>        <span class="n">mat</span><span class="p">[:,</span> <span class="n">mask_row</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
<span class="ln">16</span>        
<span class="ln">17</span>        <span class="bp">self</span><span class="o">.</span><span class="n">bias</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="ln">18</span>        <span class="bp">self</span><span class="o">.</span><span class="n">_bias_var</span> <span class="o">=</span> <span class="p">[]</span>
<span class="ln">19</span>        <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span>
<span class="ln">20</span>        <span class="n">mat_sum</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span>   <span class="c1"># force the sum of matrix to stay the same, otherwise ICE may not converge as fast</span>
<span class="ln">21</span>
<span class="ln">22</span>        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_iter</span><span class="p">):</span>
<span class="ln">23</span>            <span class="n">bias</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="ln">24</span>            <span class="n">bias_mean</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">bias</span><span class="p">[</span><span class="n">bias</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">])</span>
<span class="ln">25</span>            <span class="n">bias_var</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">bias</span><span class="p">[</span><span class="n">bias</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">])</span>
<span class="ln">26</span>            <span class="bp">self</span><span class="o">.</span><span class="n">_bias_var</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bias_var</span><span class="p">)</span>
<span class="ln">27</span>            <span class="n">bias</span> <span class="o">=</span> <span class="n">bias</span> <span class="o">/</span> <span class="n">bias_mean</span>
<span class="ln">28</span>            <span class="n">bias</span><span class="p">[</span><span class="n">bias</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
<span class="ln">29</span>
<span class="ln">30</span>            <span class="n">mat</span><span class="p">[</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">]</span> <span class="o">=</span> <span class="n">mat</span><span class="p">[</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">]</span> <span class="o">/</span> <span class="p">(</span><span class="n">bias</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">*</span><span class="n">bias</span><span class="p">[</span><span class="n">y</span><span class="p">])</span>
<span class="ln">31</span>            <span class="n">new_sum</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">mat</span><span class="p">)</span>
<span class="ln">32</span>            <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span> <span class="o">*</span> <span class="p">(</span><span class="n">mat_sum</span> <span class="o">/</span> <span class="n">new_sum</span><span class="p">)</span>
<span class="ln">33</span>
<span class="ln">34</span>            <span class="c1"># update total bias</span>
<span class="ln">35</span>            <span class="bp">self</span><span class="o">.</span><span class="n">bias</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bias</span> <span class="o">*</span> <span class="n">bias</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">new_sum</span> <span class="o">/</span> <span class="n">mat_sum</span><span class="p">)</span>
<span class="ln">36</span>        
<span class="ln">37</span>        <span class="bp">self</span><span class="o">.</span><span class="n">norm_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">matrix</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
<span class="ln">38</span>        <span class="bp">self</span><span class="o">.</span><span class="n">norm_matrix</span><span class="p">[</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">norm_matrix</span><span class="p">[</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">]</span> <span class="o">/</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">bias</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">bias</span><span class="p">[</span><span class="n">y</span><span class="p">])</span>
<span class="ln">39</span>        
<span class="ln">40</span>    <span class="k">def</span> <span class="nf">vanilla_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="ln">41</span>        <span class="bp">self</span><span class="o">.</span><span class="n">iterative_correction</span><span class="p">(</span><span class="n">max_iter</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></code></pre></div>

<p>The idea behind above algorithm is that we start by setting the biases to the sums of each row of the matrix, and dividing each matrix element by its row- and column-wise biases. The two steps are repeated until convergence criteria are met. We can use the variance of the biases (<em>self.bias</em>) to monitor the convergence of the balancing process (Figure shown below).</p>

<p><img src="/img/6.Fig1.1.png" alt="Figure 1" /></p>

<p>The raw contact matrix, matrices normalized by SP method and VC method are plotted as heatmaps as below. We can see that regions away from diagonal in the normalized matrices are cleaner than the raw matrix, but we can barely see the difference between SP and VC method. In practice, we pre-filter rows with very small values before normalization. The above script can be easily extended to achieve this. What we need to add is to set rows whose sums are smaller than certain threshold to be all zeros.</p>

<p><img src="/img/6.Fig2.1.png" alt="Figure 2" /></p>

<p>However, we can quantify the difference of SP and VC normalized methods by examining the correlation of contacts separated by the same distances. To do this, we extract and calculate the correlation of the d-th diagonal of the two matrices, where d is the distance (at bins) of two genomic regions. From the figure below, while all three methods resemble raw matrix at long distance (&gt;10 Mb), SP is slightly more similar to raw matrix. Pairwise comparison of three methods shows that SP and VC are highly similar given that they&rsquo;re only differed by number of iterations. KR resembles SP more than VC, but the difference is neglectable. [<em>Note: KR matrix is obtained from the .hic file with straw API</em>].</p>

<p><img src="/img/6.Fig3.1.png" alt="Figure" /></p>

<p>In practise, HiC normalization is done by intra-chromosome only, inter-chromosome only, or both. For intra-chromosome only, KR or ICE was done on each chromosome separately. For inter-chromosome only, the genome-wide matrix is obtained and from which the intra-chr contacts are removed. When inter-chromosome interactions are included, normalization at high resolution requires a large amount of memory. So the normalizatioln is often done on the genome-wide matrix at 25kb or 50kb resolution, as in Rao et al. A natural question is whether removing inter-chromosome contacts would have an impact on the normalization of intra-chromosome contacts? To answer this question, I performed SP normalization on All contacts vs only on intra-chromosome contacts. Again, <strong>the difference between normalized by all contacts and by only intra-chromosome contacts are really small</strong>. Juicertools and Cooler both use all contacts for normalization by default.</p>

<p><img src="/img/6.Fig4.1.png" alt="Figure" /></p>

<h3 id="conclusion">Conclusion</h3>

<p><strong>So which normalization method should I use for my data? The answer might be that it really doesn&rsquo;t matter</strong>. As shown by Rao et al and our analysis, there are high correlations observed between VC, ICE(SP), KR and a couple of other matrix balancing method. Rao et al further point out that the loop calls are not affected by different normalization method in their study. They even show that loops were nearly the same when peaks were called on raw data, which makes me question whether we need matrix normalization or not. Although Rao and others did mention that VC tends to over-correct the contact map and proposed VCSQ to supplement VC, I have not seen anyone show clear evidences that KR/ICE are superior to VC/VCSQ. The pros and cons for each method is summarized as below:
<br></p>

<table>
<thead>
<tr>
<th>Method</th>
<th>Pros</th>
<th>Cons</th>
</tr>
</thead>

<tbody>
<tr>
<td>VC/VCSQ</td>
<td>faster?</td>
<td>over-correction</td>
</tr>

<tr>
<td>ICE (SP)</td>
<td>stable?</td>
<td>slower?</td>
</tr>

<tr>
<td>KR</td>
<td>fastest? esp. for high-res maps</td>
<td>unstable for low-res maps</td>
</tr>
</tbody>
</table>

<p><br></p>

<p>In conclusion, I would still recommend normalization as part of the best pratise. Choosing KR or ICE is largely depending on the pipeline you use(Juicertools or Cooler), but if you have a relatively low sequencing depth, I would recommend ICE as it guarantees convergence.</p>

<p><br></p>

<h4 id="code-availability">Code availability</h4>

<p>All codes used for this post can be found: <a href="https://github.com/dearxxj/genomics_tools/blob/master/HiC_Normalization/HiC-normalization.ipynb">https://github.com/dearxxj/genomics_tools/blob/master/HiC_Normalization/HiC-normalization.ipynb</a>
<br></p>

<p>Reference:<br>
<strong>[1]</strong>. Imakaev et al. Iterative correction of hi-c data reveals hallmarks of chromosome organization. Nature Method. 2012. <br>
<strong>[2]</strong>. Rao et al. A 3D Map of the Human Genome at Kilobase Resolution Reveals Principles of Chromatin Looping. Cell. 2014.</p>


        
          <div class="blog-tags">
            
              <a href="https://dearxxj.github.io/tags/bioinfo/">Bioinfo</a>&nbsp;
            
              <a href="https://dearxxj.github.io/tags/hi-c/">Hi-C</a>&nbsp;
            
          </div>
        

        
            <hr/>
            <section id="social-share">
              <div class="list-inline footer-links">
                

<div class="share-box" aria-hidden="true">
    <ul class="share">
      
      <li>
        <a href="//twitter.com/share?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f6%2f&amp;text=Demystify%20Hi-C%20Data%20Normalization&amp;via=" target="_blank" title="Share on Twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.facebook.com/sharer/sharer.php?u=https%3a%2f%2fdearxxj.github.io%2fpost%2f6%2f" target="_blank" title="Share on Facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//reddit.com/submit?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f6%2f&amp;title=Demystify%20Hi-C%20Data%20Normalization" target="_blank" title="Share on Reddit">
          <i class="fab fa-reddit"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.linkedin.com/shareArticle?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f6%2f&amp;title=Demystify%20Hi-C%20Data%20Normalization" target="_blank" title="Share on LinkedIn">
          <i class="fab fa-linkedin"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.stumbleupon.com/submit?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f6%2f&amp;title=Demystify%20Hi-C%20Data%20Normalization" target="_blank" title="Share on StumbleUpon">
          <i class="fab fa-stumbleupon"></i>
        </a>
      </li>
  
      
      <li>
        <a href="//www.pinterest.com/pin/create/button/?url=https%3a%2f%2fdearxxj.github.io%2fpost%2f6%2f&amp;description=Demystify%20Hi-C%20Data%20Normalization" target="_blank" title="Share on Pinterest">
          <i class="fab fa-pinterest"></i>
        </a>
      </li>
    </ul>
  </div>
  

              </div>
            </section>
        

        
          
            
          

          
                  <h4 class="see-also">See also</h4>
                  <ul>
                
                
                    <li><a href="/post/5/">Learning HTSlib (1)</a></li>
                
                    <li><a href="/post/4/">Exact string matching algorithms: Boyer-Moore and KMP</a></li>
                
                    <li><a href="/post/3/">Similarity measurement of two sets of TADs</a></li>
                
                    <li><a href="/post/2/">How does number of threads affect mapping time in BWA MEM</a></li>
                
              </ul>

          
        
      </article>

      
        <ul class="pager blog-pager">
          
            <li class="previous">
              <a href="https://dearxxj.github.io/post/5/" data-toggle="tooltip" data-placement="top" title="Learning HTSlib (1)">&larr; Previous Post</a>
            </li>
          
          
        </ul>
      


      
        
          
          <div class="disqus-comments">
            <div id="disqus_thread"></div>
<script type="application/javascript">
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "fsong" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
          </div>
          
        
        
      

    </div>
  </div>
</div>

      
<footer>
  <div class="container">
    <div class="row">
      <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
        <ul class="list-inline text-center footer-links">
          
              <li>
                <a href="https://github.com/dearxxj" title="GitHub">
                  <span class="fa-stack fa-lg">
                    <i class="fas fa-circle fa-stack-2x"></i>
                    <i class="fab fa-github fa-stack-1x fa-inverse"></i>
                  </span>
                </a>
              </li>
          
          <li>
            <a href="" title="RSS">
              <span class="fa-stack fa-lg">
                <i class="fas fa-circle fa-stack-2x"></i>
                <i class="fas fa-rss fa-stack-1x fa-inverse"></i>
              </span>
            </a>
          </li>
          
        </ul>
        <p class="credits copyright text-muted">
          
            
              <a href="https://dearxxj.github.io">Fan</a>
            
          

          &nbsp;&bull;&nbsp;&copy;
          
            2019
          

          
            &nbsp;&bull;&nbsp;
            <a href="https://dearxxj.github.io">Fan&#39;s Blog</a>
          
        </p>
        
        <p class="credits theme-by text-muted">
          <a href="https://gohugo.io">Hugo v0.59.1</a> powered &nbsp;&bull;&nbsp; Theme <a href="https://github.com/halogenica/beautifulhugo">Beautiful Hugo</a> adapted from <a href="https://deanattali.com/beautiful-jekyll/">Beautiful Jekyll</a>
          
        </p>
      </div>
    </div>
  </div>
</footer><script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.js" integrity="sha384-K3vbOmF2BtaVai+Qk37uypf7VrgBubhQreNQe9aGsz9lB63dIFiQVlJbr92dw2Lx" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/contrib/auto-render.min.js" integrity="sha384-kmZOZB5ObwgQnS/DuDg6TScgOiWWBiVt0plIRkZCmE6rDZGrEOQeHM5PcHi+nyqe" crossorigin="anonymous"></script>
<script src="https://code.jquery.com/jquery-1.12.4.min.js" integrity="sha256-ZosEbRLbNQzLpnKIkEdrPv7lOy9C27hHQ+Xp8a4MxAQ=" crossorigin="anonymous"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<script src="https://dearxxj.github.io/js/main.js"></script>
<script src="https://dearxxj.github.io/js/highlight.min.js"></script>
<script> hljs.initHighlightingOnLoad(); </script>
<script> $(document).ready(function() {$("pre.chroma").css("padding","0");}); </script><script> renderMathInElement(document.body); </script><script src="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe.min.js" integrity="sha384-QELNnmcmU8IR9ZAykt67vGr9/rZJdHbiWi64V88fCPaOohUlHCqUD/unNN0BXSqy" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe-ui-default.min.js" integrity="sha384-m67o7SkQ1ALzKZIFh4CiTA8tmadaujiTa9Vu+nqPSwDOqHrDmxLezTdFln8077+q" crossorigin="anonymous"></script><script src="https://dearxxj.github.io/js/load-photoswipe.js"></script>









    
  </body>
</html>

